dali/util/half.h

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_UTIL_HALF_H_
+#define DALI_UTIL_HALF_H_
+
+// Macros for writing RunImpl templated code for derived operators
+template <typename Out>
+class CastOut {
+ public:
+  inline Out operator()(float arg)        { return static_cast<Out>(arg); }
+};
+
+#ifndef __CUDA_ARCH__
+#ifdef DALI_F16C
+  #include <x86intrin.h>
+
+class CastF16C {
+ public:
+  inline uint16_t operator()(float arg)   { return _cvtss_sh(arg, 0); }
+};
+
+  #define RUN_IMPL_CPU(ws, idx)  RUN_IMPLEMENT(ws, idx, uint16_t, CastF16C)
+
+#else
+  #include "dali/util/half.hpp"
+
+  #define RUN_IMPL_CPU(ws, idx)  RUN_IMPL(ws, idx, half_float::half)
+
+#endif
+#endif  //  __CUDA_ARCH__
+
+#define RUN_IMPL_GPU(ws, idx)    RUN_IMPL(ws, idx, float16)
+
+#define RUN_IMPLEMENT(ws, idx, out, cast)       \
```

Comment:
Specializing CastOut will allow to get rid of macros for invocation of kernels.

Reply:
As I explained in my email, it makes no sense to get rid of these macros. But I significantly simplified them and from now only one of them will be used.